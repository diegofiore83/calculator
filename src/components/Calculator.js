import React, { useState } from 'react';
import { Box, Button, Paper } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import { calculator } from '../modules/calculator';
import logo from '../assets/logo.png';

const boxProps = className => ({ className, boxShadow: 2, bgcolor: 'background.paper' });
const buttonProps = className => ({ className, color: 'secondary', variant: 'contained' });

const useStyles = makeStyles(theme => ({
  calculator: {
    backgroundColor: theme.palette.primary.light,
    maxWidth: theme.spacing(32),
  },
  calculatorButton: {
    fontSize: 24,
    fontWeight: 'bold',
    margin: theme.spacing(1),
    padding: 0,
    minHeight: theme.spacing(6),
    minWidth: theme.spacing(6),
    maxHeight: theme.spacing(6),
    maxWidth: theme.spacing(6),
  },
  calculatorScreen: {
    fontSize: 24,
    fontWeight: 'bold',
    textAlign: 'right',
    borderRadius: 4,
    margin: theme.spacing(1),
    padding: theme.spacing(1),
    width: `calc(100% - ${theme.spacing(2)}px)`,
    overflow: 'hidden',
  },
  calculatorLogo: {
    width: '100%', 
    pointerEvents: 'none',
  }
}));

const Component = () => {
  const [operand1, setOperand1] = useState('0');
  const [operand2, setOperand2] = useState(null);
  const [operator, setOperator] = useState(null);

  const classes = useStyles();

  const operandClick = value => {
    if (operator) {
      if (operand2 === null) {
        setOperand2(value);
      } else {
        setOperand2(operand2 + value);
      }
    } else {
      if (operand1 === '0') {
        setOperand1(value);
      } else {
        setOperand1(operand1 + value);
      }
    }
  };

  const operatorClick = value => {
    if (operand2 === null) {
      if (value !== '=') {
        setOperator(value);
      }
    } else {
      setOperand1(calculator({ operand1: parseInt(operand1), operator, operand2: parseInt(operand2) }));
      setOperator(value === '=' ? null : value);
      setOperand2(null);
    }
  };

  const resetCalculator = () => {
    setOperand1('0');
    setOperator(null);
    setOperand2(null);
  };

  return (
    <Paper className={classes.calculator}>
      <Box id="screen" {...boxProps(classes.calculatorScreen)}>{operand2 || operand1}</Box>
      <Button id="operand7" {...buttonProps(classes.calculatorButton)} onClick={() => operandClick('7')}>7</Button>
      <Button id="operand8" {...buttonProps(classes.calculatorButton)} onClick={() => operandClick('8')}>8</Button>
      <Button id="operand9" {...buttonProps(classes.calculatorButton)} onClick={() => operandClick('9')}>9</Button>
      <Button id="operatorAddition" {...buttonProps(classes.calculatorButton)} onClick={() => operatorClick('+')}>+</Button>
      <Button id="operand4" {...buttonProps(classes.calculatorButton)} onClick={() => operandClick('4')}>4</Button>
      <Button id="operand5" {...buttonProps(classes.calculatorButton)} onClick={() => operandClick('5')}>5</Button>
      <Button id="operand6" {...buttonProps(classes.calculatorButton)} onClick={() => operandClick('6')}>6</Button>
      <Button id="operatorSubtraction" {...buttonProps(classes.calculatorButton)} onClick={() => operatorClick('-')}>-</Button>
      <Button id="operand1" {...buttonProps(classes.calculatorButton)} onClick={() => operandClick('1')}>1</Button>
      <Button id="operand2" {...buttonProps(classes.calculatorButton)} onClick={() => operandClick('2')}>2</Button>
      <Button id="operand3" {...buttonProps(classes.calculatorButton)} onClick={() => operandClick('3')}>3</Button>
      <Button id="operatorMultiplication" {...buttonProps(classes.calculatorButton)} onClick={() => operatorClick('*')}>*</Button>
      <Button id="operatorCancel" {...buttonProps(classes.calculatorButton)} onClick={() => resetCalculator()}>C</Button>
      <Button id="operand0" {...buttonProps(classes.calculatorButton)} onClick={() => operandClick('0')}>0</Button>
      <Button id="operatorEqual" {...buttonProps(classes.calculatorButton)} onClick={() => operatorClick('=')}>=</Button>
      <Button id="operatorDivision" {...buttonProps(classes.calculatorButton)} onClick={() => operatorClick('/')}>/</Button>
      <img src={logo} className={classes.calculatorLogo} alt="logo" />
    </Paper>
  );
};

Component.displayName = 'Calculator';

Component.propTypes = {};

Component.defaultProps = {};

export default Component;
