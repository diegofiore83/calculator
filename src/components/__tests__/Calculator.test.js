import React from 'react';
import { shallow } from 'enzyme';
import Calculator from '../Calculator';

describe('Calculator', () => {
  const props = {};

  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, 'useState');
  useStateSpy.mockImplementation(init => [init, setState]);

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Calculator {...props} />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('render', () => {
    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('behaviour', () => {
    it('should set the first operand', () => {
      wrapper.find('#operand6').first().simulate('click');

      expect(wrapper.find('#screen').text()).toBe('6');
    });

    it('should not change operand when 0 is pressed twice or more', () => {
      wrapper.find('#operand0').first().simulate('click');
      wrapper.find('#operand0').first().simulate('click');

      expect(wrapper.find('#screen').text()).toBe('0');
    });

    it('should set the first operand second digit', () => {
      wrapper.find('#operand6').first().simulate('click');
      wrapper.find('#operand7').first().simulate('click');

      expect(wrapper.find('#screen').text()).toBe('67');
    });

    it('should set the first operand third digit', () => {
      wrapper.find('#operand6').first().simulate('click');
      wrapper.find('#operand7').first().simulate('click');
      wrapper.find('#operand0').first().simulate('click');

      expect(wrapper.find('#screen').text()).toBe('670');
    });

    it('should set the first three digit on the screen after equal selected', () => {
      wrapper.find('#operand1').first().simulate('click');
      wrapper.find('#operand5').first().simulate('click');
      wrapper.find('#operand2').first().simulate('click');
      wrapper.find('#operatorEqual').first().simulate('click');
      wrapper.find('#operand0').first().simulate('click');

      expect(wrapper.find('#screen').text()).toBe('1520');
    });

    it('should set the second operator on the screen after addition selected', () => {
      wrapper.find('#operand1').first().simulate('click');
      wrapper.find('#operand5').first().simulate('click');
      wrapper.find('#operand2').first().simulate('click');
      wrapper.find('#operatorAddition').first().simulate('click');
      wrapper.find('#operand1').first().simulate('click');
      wrapper.find('#operand0').first().simulate('click');

      expect(wrapper.find('#screen').text()).toBe('10');
    });

    it('should set the second operator on the screen after equal selected', () => {
      wrapper.find('#operand1').first().simulate('click');
      wrapper.find('#operand5').first().simulate('click');
      wrapper.find('#operand2').first().simulate('click');
      wrapper.find('#operatorAddition').first().simulate('click');
      wrapper.find('#operand1').first().simulate('click');
      wrapper.find('#operand0').first().simulate('click');
      wrapper.find('#operatorEqual').first().simulate('click');

      expect(wrapper.find('#screen').text()).toBe('162');
    });

    it('should do multiple operation in a row', () => {
        wrapper.find('#operand9').first().simulate('click');
        wrapper.find('#operatorMultiplication').first().simulate('click');
        wrapper.find('#operand4').first().simulate('click');
        wrapper.find('#operatorDivision').first().simulate('click');
        wrapper.find('#operand3').first().simulate('click');
        wrapper.find('#operatorSubtraction').first().simulate('click');
        wrapper.find('#operand8').first().simulate('click');
        wrapper.find('#operatorEqual').first().simulate('click');
  
        expect(wrapper.find('#screen').text()).toBe('4');
      });

      it('should cancel the value', () => {
        wrapper.find('#operand9').first().simulate('click');
        wrapper.find('#operand4').first().simulate('click');
        wrapper.find('#operatorCancel').first().simulate('click');
  
        expect(wrapper.find('#screen').text()).toBe('0');
      });
  });
});
