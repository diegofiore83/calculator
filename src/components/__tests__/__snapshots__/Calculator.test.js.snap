// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator render should render correctly 1`] = `
<WithStyles(ForwardRef(Paper))
  className="makeStyles-calculator-1"
>
  <Styled(MuiBox)
    bgcolor="background.paper"
    boxShadow={2}
    className="makeStyles-calculatorScreen-3"
    id="screen"
  >
    0
  </Styled(MuiBox)>
  <WithStyles(ForwardRef(Button))
    className="makeStyles-calculatorButton-2"
    color="secondary"
    id="operand7"
    onClick={[Function]}
    variant="contained"
  >
    7
  </WithStyles(ForwardRef(Button))>
  <WithStyles(ForwardRef(Button))
    className="makeStyles-calculatorButton-2"
    color="secondary"
    id="operand8"
    onClick={[Function]}
    variant="contained"
  >
    8
  </WithStyles(ForwardRef(Button))>
  <WithStyles(ForwardRef(Button))
    className="makeStyles-calculatorButton-2"
    color="secondary"
    id="operand9"
    onClick={[Function]}
    variant="contained"
  >
    9
  </WithStyles(ForwardRef(Button))>
  <WithStyles(ForwardRef(Button))
    className="makeStyles-calculatorButton-2"
    color="secondary"
    id="operatorAddition"
    onClick={[Function]}
    variant="contained"
  >
    +
  </WithStyles(ForwardRef(Button))>
  <WithStyles(ForwardRef(Button))
    className="makeStyles-calculatorButton-2"
    color="secondary"
    id="operand4"
    onClick={[Function]}
    variant="contained"
  >
    4
  </WithStyles(ForwardRef(Button))>
  <WithStyles(ForwardRef(Button))
    className="makeStyles-calculatorButton-2"
    color="secondary"
    id="operand5"
    onClick={[Function]}
    variant="contained"
  >
    5
  </WithStyles(ForwardRef(Button))>
  <WithStyles(ForwardRef(Button))
    className="makeStyles-calculatorButton-2"
    color="secondary"
    id="operand6"
    onClick={[Function]}
    variant="contained"
  >
    6
  </WithStyles(ForwardRef(Button))>
  <WithStyles(ForwardRef(Button))
    className="makeStyles-calculatorButton-2"
    color="secondary"
    id="operatorSubtraction"
    onClick={[Function]}
    variant="contained"
  >
    -
  </WithStyles(ForwardRef(Button))>
  <WithStyles(ForwardRef(Button))
    className="makeStyles-calculatorButton-2"
    color="secondary"
    id="operand1"
    onClick={[Function]}
    variant="contained"
  >
    1
  </WithStyles(ForwardRef(Button))>
  <WithStyles(ForwardRef(Button))
    className="makeStyles-calculatorButton-2"
    color="secondary"
    id="operand2"
    onClick={[Function]}
    variant="contained"
  >
    2
  </WithStyles(ForwardRef(Button))>
  <WithStyles(ForwardRef(Button))
    className="makeStyles-calculatorButton-2"
    color="secondary"
    id="operand3"
    onClick={[Function]}
    variant="contained"
  >
    3
  </WithStyles(ForwardRef(Button))>
  <WithStyles(ForwardRef(Button))
    className="makeStyles-calculatorButton-2"
    color="secondary"
    id="operatorMultiplication"
    onClick={[Function]}
    variant="contained"
  >
    *
  </WithStyles(ForwardRef(Button))>
  <WithStyles(ForwardRef(Button))
    className="makeStyles-calculatorButton-2"
    color="secondary"
    id="operatorCancel"
    onClick={[Function]}
    variant="contained"
  >
    C
  </WithStyles(ForwardRef(Button))>
  <WithStyles(ForwardRef(Button))
    className="makeStyles-calculatorButton-2"
    color="secondary"
    id="operand0"
    onClick={[Function]}
    variant="contained"
  >
    0
  </WithStyles(ForwardRef(Button))>
  <WithStyles(ForwardRef(Button))
    className="makeStyles-calculatorButton-2"
    color="secondary"
    id="operatorEqual"
    onClick={[Function]}
    variant="contained"
  >
    =
  </WithStyles(ForwardRef(Button))>
  <WithStyles(ForwardRef(Button))
    className="makeStyles-calculatorButton-2"
    color="secondary"
    id="operatorDivision"
    onClick={[Function]}
    variant="contained"
  >
    /
  </WithStyles(ForwardRef(Button))>
  <img
    alt="logo"
    className="makeStyles-calculatorLogo-4"
    src="logo.png"
  />
</WithStyles(ForwardRef(Paper))>
`;
